docs/architecture/blazor-for-web-forms-developers/components.md

```diff
@@ -3,11 +3,540 @@ title: Build reusable UI components with Blazor
+
+```razor
+<button @onclick="OnClick">Click me!</button>
+
+@code {
+    void OnClick() 
+    {
+        Console.WriteLine("The button was clicked!);
+    }
+}
+```
+
+Event handlers can take an optional event-specific argument to provide more information about the event. For example, mouse events can take a `MouseEventArgs` argument, but it isn't required.
+
+```razor
+<button @onclick="OnClick">Click me!</button>
+
+@code {
+    void OnClick(MouseEventArgs e) 
+    {
+        Console.WriteLine($"Mouse clicked at {e.ScreenX}, {e.ScreenY}.");
+    }
+}
+```
+
+Instead of referring to a method group for an event handler, you can use a lambda expression. A lambda expression allows you to close over other in-scope values.
+
+```razor
+@foreach (var buttonLabel in buttonLabels)
+{
+    <button @onclick="() => Console.WriteLine($"The {buttonLabel} button was clicked!")">@buttonLabel</button>
+}
+```
+
+Event handlers can execute synchronously or asynchronously.
+
+```razor
+<button @onclick="OnClick">Click me!</button>
+
+@code {
+    async Task OnClick() 
+    {
+        var result = await Http.GetAsync("api/values");
+    }
+}
+```
+
+After an event is handled, the component is rendered to account for any component state changes. With asynchronous event handlers, the component is rendered immediately after the handler execution completes. The component is rendered *again* after the asynchronous `Task` completes. This asynchronous execution mode provides an opportunity to render some appropriate UI while the asynchronous `Task` is still in progress.
+
+```razor
+<button @onclick="Get message">Get message</button>
+
+@if (showMessage)
+{
+    @if (message == null)
+    {
+        <p><em>Loading...</em></p>
+    }
+    else
+    {
+        <p>The message is: @message</p>
+    }
+}
+
+@code 
+{
+    bool showMessage = false;
+    string message;
+
+    public async Task ShowMessage()
+    {
+        showMessage = true;
+        message = await MessageService.GetMessageAsync();
+    }
+}
+```
+
+Components can also define their own events by defining a component parameter of type `EventCallback<TValue>`. Event callbacks support all the variations of DOM UI event handlers: optional args, sync or async, method groups or lambda expressions.
+
+```razor
+<button class="btn btn-primary" @onclick="OnClick">Click me!</button>
+
+@code {
+    [Parameter]
+    public EventCallback<MouseEventArgs> OnClick { get; set; }
+}
+```
+
+## Data binding
+
+Blazor provides a simple mechanism to bind data from a UI component to the component's state. This approach differs from the features in ASP.NET Web Forms for binding data from data sources to UI controls. We'll cover handling data from different data sources in the [Dealing with data](data.md) section.
+
+To create a two-way data binding from a UI component to some the component's state, use the `@bind` directive attribute. In the example below, the value of the check box is bound to the `isChecked` field.
```

Comment:
Something is off in this first sentence. Is there a word missing between "some" and "the"?